pyfakefs/tests/fake_filesystem_test.py

```diff
@@ -852,6 +853,9 @@ def tearDown(self):
 
     def test_create_top_level_directory(self):
         top_level_dir = "/x"
+        # remove the existing temp dir
```

Comment:
Is this because /tmp is not expected further down? This seems to be forcing the test function to conform to the new behavior.

Reply:
True.
First - thanks for the review. It is good to have someone else to check the code again...

For a bit of background: 
Previously (that was several years ago), the `tempfile` module was patched in `pyfakefs`, but the patch was incomplete and had problems, and broke with every new Python version. So I removed it from patching (together with `glob` which had a similar problem) and solely relied on the patched `os` / `os.path` functionality. That worked, except that `tempfile` assumes an existing temp directory, otherwise it crashes, so I decided to create that directory on startup.

It is true that this changes the behavior [if you expected an empty filesystem](https://pytest-pyfakefs.readthedocs.io/en/latest/troubleshooting.html#os-temporary-directories), but so far I had no reports that this causes big problems.

I was also not sure about the usage pattern of `pyfakefs`. Recently I did some statistics with the repositories shown as dependent on `pyfakefs` (as this is the only source I have for that), counting only repositories that had at least 100 stars (about 50, so not really a large sample), and found that of these, more than 40% used it via `pytest`, and about the same number via the `unittest` support added by @jmcgeheeiv. Another 5% used the patcher directly, and the rest (4 in this case) had been manually patching as originally used by Google.
If assume that the usage you are referring to is inside Google, which is good to know. 

More to the point: I could add an argument to the fake filesystem initialization (defaulting to `False`) that would control if the temp file shall be created. I would probably set that to `True` if initialized from the patcher - this way, it would retain the current behavior if invoking `pyfakefs` via `pytest` or `unittest`, and would also retain the behavior if creating it manually. Would that be ok for you?

Reply:
Generally, I think that we have to keep 2 usage scenarios in mind: replacing the whole fs using the patcher (directly or via the `fs` fixture, `fake_filesystem_unittest.TestCase` or `@patchfs`), and the usage by directly instantiating the fake filesystem and the fake modules, where needed, possibly subclassing them. The first use case is what I mostly had in mind so far, and that is mostly used by people I get feedback from. The second scenario is the one originally used (and obviously still used) at Google and certainly elsewhere, but so far we had no feedback from these users.

Reply:
Yes, there were previously three additional modules that were removed, `fake_tempfile`, `fake_filesystem_glob` and `fake_filesystem_shutil`. When mocking modules with `fake_filesystem` through `unittest.mock`, which is common in the Google unit tests, the unit tests would add mocking behavior to `tempfile` through the `fake_tempfile` module. The other two modules are simple to mock, `shutil` and `glob` by mocking their `os` imports directly. However, `tempfile` is, as you note, very complex, and so after `fake_tempfile` was removed, those unit tests ported their implementations to `fake_filesystem_unittest` which provides a complete implementation of a POSIX file system.

In the case of adding `fake_filesyste_unittest` to unit tests, I think it is very appropriate to add the `tempfile.gettempdir` directory for POSIX file system support. I also think that resetting the file system for each test case makes a lot of sense. In fact, I encourage use of the `@fake_filesystem_unittest.patchfs` decorator instead of subclassing `fake_filessytem_unittest.TestCase`.

I think that for `fake_filesystem` however, the only expectation is a clean fake filesystem, and nothing more. In fact, internally at Google, we have other file systems other than the local filesystem that we provide fakes for, that extend pyfakefs fake filesystem, and for which there are no devices (no /dev/null that was added at some point to pyfakefs) and no /tmp directory, because they don't make sense in those situations.

It is very simple to add a temporary directory when setting up a test function to a test module that imports `fake_filesystem` and mocks the imports of modules under test, so I would prefer preserving that expectation.

I may need to add documentation on usage of `fake_filesystem` with the `unittest.mock` module, and also, how to extend `fake_filesystem` for non-POSIX, Windows or MacOS file systems, that are not patching imports of the modules under test, and are rather, accessing those file systems through other APIs.

Thanks.

Reply:
Thanks, this all makes sense. As I wrote, that is a different use case, that I hadn't really looked upon - I myself started using `pyfakefs` in a django environment under Windows (don't ask), and had always used `fake_filesystem_unittest.TestCase`, or later the `fs` fixture. The first thing I added was better Windows support, and a couple of features (like support for file system size) that I needed myself. Afterwards the development was mostly driven by issues and feature requests, and by new Python versions (other than infrastructure stuff like documentation, CI and more tests), and none of these so far was related to the usage with `unittest.mock` (it isn't even in the documentation). Hopefully this will now change...

As for this PR - I understand it is ok for you in the current incarnation (e.g. with the previous behavior in place)? I'll wait for your approval before merging.

Reply:
Is it possible to address the regression when resetting the `fs` fixture without adding tempfile support to the `FakeFilsystem` class by default? I understand the intent of this code change, but since it is responding to a regression of `fs` fixtures, I think it should be limited to resetting `fs` fixtures.

This change code change will impact all clients of `FileSystem`. I'm pretty sure that many of the unit tests I support (20,000 at latest count) will fail because of assumptions about the file system not containing a `/tmp` directory. The change is well intentioned, and perhaps a different implementation could provide an initial directory structure to `FakeFilesystem` through a new `__init__` arg. This could be useful for many other clients of `FakeFilesystem` with their own initial file directory structure.

In terms of `pyfakefs` design, I'd like to review recent changes, and as I said, add explanations of using the `FakeFilesystem` class for other file system interfaces not supported by the `fs` fixture. As I said, it is used by special file systems. I worked on the capability to fake a Windows file system from a unit test running on Linux many many years ago (2012?), and I am glad that in general, it is a useful feature and has continued support.

Reply:
As I wrote, after the change I made after your comment, tempfile creation is switched off by default. It is only switched on if created from the patcher (or if you explicitely switch it on), so the behavior should not change for the existing clients. Maybe you hadn't noticed the commit, as I just amended the old one...
Leaving the code in the patcher would make it difficult to call it on filesystem reset.

About the review: I will be glad if you can review the code and suggest changes. I understand you want to add the explanations to the `pyfakefs` documentation and not to your internal documentation? That would indeed by helpful!

Reply:
Oh... sorry... I didn't see the update in the source files. My bad. I'll continue my code review there.